src/partition.ml

```diff
@@ -0,0 +1,94 @@
+open Encoding
+module H_contents = Irmin.Hash.Typed (Hash) (Contents)
+
+type entry = { name : string; kind : [ `Node | `Contents ]; hash : Hash.t }
+type enc_entry = { eencoding : string; entry : entry }
+type pointer = { index : int; hash : Hash.t }
+type enc_pointer = { pencoding : string; pointer : pointer }
+
+type vs =
+  | Empty
+  | Values of enc_entry list
+  | Tree of {
+      depth : int;
+      length : int;
+      pointers : (enc_pointer * enc_vs) list;
+    }
+
+and enc_vs = { vsencoding : string; vs : vs }
+
+let pre_hash t v =
+  let buf = Buffer.create 13 in
+  let pre_hash = Irmin.Type.(unstage (pre_hash t)) in
+  pre_hash v (Buffer.add_string buf);
+  Buffer.contents buf
+
+let compute_hash s = H_contents.hash (Bytes.of_string s)
+let pp_hash fmt h = Irmin.Type.(pp Encoding.Hash.t) fmt h
+let char_of_kind = function `Contents -> '\000' | `Node -> '\001'
+
+let to_enc_entry ({ name; kind; hash } as entry) =
+  let len = pre_hash Irmin.Type.int (String.length name) in
+  let eencoding =
+    Format.asprintf "%s%s%c%a" len name (char_of_kind kind) pp_hash hash
```

Comment:
your problem might come from here -  the `pp_hash` introduces the "Co" prefix in the encoding of an entry

Reply:
Ah yes, maybe! Is there a way to not have it?

Reply:
you can use `to_bin_string` to serialize the hash to a string (and yes you definitely don't want to pretty-print it!)

Reply:
I'm not sure using `Format` here is the most simple way to do this. It would be nice to have an encoding function which looks like the spec, e.g. for Inode values something like:

```
(** See https://hackmd.io/@samoht/tezos-context-hash#Inode-values *)
let tree_entry (n, k, h) = leb128 (String.length n) ++ n ++ kind k ++ hash h
```

Reply:
You'd rather have a function that gives the encoding or the encoding directly in the type?

Reply:
It's now implemented in https://github.com/tarides/tezos-context-hash/pull/12/files#diff-79875b9b2bc8ccf7b8414b177d2314571ef9557df7434b62b8275cb816256796R108-R112